include/gcodeExport.h

```diff
@@ -113,6 +115,22 @@ class GCodeExport : public NoCopy
         {
         }
     };
+
+    struct RetractionAmounts
+    {
+        double old_e{ 0.0 }; // The previous absolute retraction amount
+        double new_e{ 0.0 }; // The new absolute retraction amount
+        double diff_e{ 0.0 }; // The difference between the new and previous amount, which is to be processed
+
+        /*!
+         * Indicates whether this retraction actually has something to process
+         */
+        inline bool has_retraction() const
+        {
+            return std::abs(diff_e) >= 0.000001;
+        }
```

Comment:
`nodiscard` is your friend and member functions defined inside a class definition are implicitly inline.The primary improvements come from making the function `constexpr` and `noexcept`.

```suggestion
        [[nodiscard]] constexpr bool has_retraction() const noexcept
        {
            constexpr double threshold { 1e-6 };
            return std::abs(diff_e) >= threshold;
        }
```

Reply:
Didn't you pay attention `constexpr` everything! https://www.youtube.com/watch?v=PJwd4JLYJJY

Reply:
But yes in this case it is still depend on a dynamic value